File: java/mojo-parser/src/main/java/org/mojolang/mojo/parser/TypeIdentifierVisitor.java
package org.mojolang.mojo.parser;

import com.google.protobuf.GeneratedMessageV3;
import org.mojolang.mojo.lang.NominalType;

import java.util.List;

public class TypeIdentifierVisitor extends MojoBaseVisitor<com.google.protobuf.GeneratedMessageV3> {
    private NominalType.Builder builder_ = null;

    @Override
    public GeneratedMessageV3 visitType_identifier(MojoParser.Type_identifierContext ctx) {

        List<MojoParser.Type_identifier_clauseContext> typeCtxList = ctx.type_identifier_clause();
        for (MojoParser.Type_identifier_clauseContext typeCtx : typeCtxList) {
            NominalType.Builder builder = visitTypeIdentifierClause(typeCtx);

            if (builder_ != null) {
                builder.setEnclosingType(builder_.build());
            }
            builder_ = builder;
        }

        return builder_.build();
    }

    private NominalType.Builder visitTypeIdentifierClause(MojoParser.Type_identifier_clauseContext ctx) {
        NominalType.Builder builder = NominalType.newBuilder();

        builder.setName(ctx.type_name().getText());

        if (ctx.generic_argument_clause() != null) {
            builder.addAllGenericArguments(Util.visitGenericArguments(ctx.generic_argument_clause()));
        }

        return builder;
    }
}

File: go/pkg/parser/type_identifier_visitor.go
package parser

import "github.com/mojo-lang/lang/go/pkg/lang"

type TypeIdentifierVisitor struct {
	*BaseMojoParserVisitor
}

func NewTypeIdentifierVisitor() *TypeIdentifierVisitor {
	visitor := &TypeIdentifierVisitor{}
	return visitor
}

func (t *TypeIdentifierVisitor) VisitTypeIdentifier(ctx *TypeIdentifierContext) interface{} {
	identifiers := ctx.AllTypeIdentifierClause()
	var nominalType *lang.NominalType

	for _, identifier := range identifiers {
		t := identifier.Accept(t).(*lang.NominalType)

		if nominalType == nil {
			nominalType = t
		} else {
			t.EnclosingType = nominalType
			nominalType = t
		}
	}

	return nominalType
}

func (t *TypeIdentifierVisitor) VisitTypeIdentifierClause(ctx *TypeIdentifierClauseContext) interface{} {
	nominalType := &lang.NominalType{}

	nameCtx := ctx.TypeName()
	if nameCtx != nil {
		visitor := NewTypeNameVisitor()
		nominalType.Name = nameCtx.Accept(visitor).(string)
	}

	genericArgumentCtx := ctx.GenericArgumentClause()
	if genericArgumentCtx != nil {
		visitor := NewGenericArgumentsVisitor()
		nominalType.GenericArguments = genericArgumentCtx.Accept(visitor).([]*lang.GenericArgument)
	}

	return nominalType
}
